engine/table/src/main/java/io/deephaven/engine/table/impl/naturaljoin/typed/incopen/gen/IncrementalNaturalJoinHasherChar.java

```diff
@@ -93,16 +102,25 @@ protected void buildFromRightSide(RowSequence rowSequence, Chunk[] sourceKeyChun
             final int hash = hash(k0);
             final int firstTableLocation = hashToTableLocation(hash);
             int tableLocation = firstTableLocation;
+            int firstDeletedLocation = -1;
             MAIN_SEARCH: while (true) {
                 long existingRightRowKey = mainRightRowKey.getUnsafe(tableLocation);
-                if (existingRightRowKey == EMPTY_RIGHT_STATE) {
-                    numEntries++;
+                if (firstDeletedLocation < 0 && isStateDeleted(existingRightRowKey)) {
+                    firstDeletedLocation = tableLocation;
+                }
+                if (isStateEmpty(existingRightRowKey)) {
+                    if (firstDeletedLocation >= 0) {
+                        tableLocation = firstDeletedLocation;
+                    } else {
+                        numEntries++;
+                    }
+                    liveEntries++;
                     mainKeySource0.set(tableLocation, k0);
                     mainLeftRowSet.set(tableLocation, RowSetFactory.empty());
                     mainRightRowKey.set(tableLocation, rowKeyChunk.get(chunkPosition));
                     mainModifiedTrackerCookieSource.set(tableLocation, -1L);
                     break;
-                } else if (eq(mainKeySource0.getUnsafe(tableLocation), k0)) {
+                } else if (!isStateDeleted(existingRightRowKey) && eq(mainKeySource0.getUnsafe(tableLocation), k0)) {
                     if (existingRightRowKey == RowSet.NULL_ROW_KEY) {
                         mainRightRowKey.set(tableLocation, rowKeyChunk.get(chunkPosition));
                     } else if (existingRightRowKey < RowSet.NULL_ROW_KEY) {
```

Comment:
This should be <= FIRST_DUPLICATE. You have this problem in a few places, only commenting here.